### PR TITLE
Jetpack Error UX: Add component that checks Jetpack connection health

### PR DESCRIPTION
**Pull request by wojtekn**
Related to https://github.com/Automattic/dotcom-forge/issues/3202

## Proposed Changes

In this diff, I propose adding a new component that checks Jetpack connection health and displays an accurate message that can let the user fix the connection.

The component should only execute a request to the API endpoint if we suspect the connection can be broken. The connection can be broken if another request made to an external site via the proxy failed.

## Testing Instructions

1. Apply patch D116410-code
2. Load the posts page for a blog with a broken connection
3. Confirm that the message is displayed
4. Confirm that there is a track logged for notice view and notice action click

Screenshot 2023-08-02 at 10 26 22

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/components/jetpack/connection-health/index.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		} );
+
+	if (
+		isLoadingJetpackConnectionHealth ||
+		isErrorCheckJetpackConnectionHealth ||
+		jetpackConnectionHealth?.is_healthy
+	) {
+		return;
+	}
+
```
Comment (sejas) [inline client/components/jetpack/connection-health/index.tsx:44]: At this point, can we determine if the jetpack connection is failing/working just by checking the API response?

Comment (wojtekn) [reply client/components/jetpack/connection-health/index.tsx:44]: If `is_healthy` is `true` it means API is ok and we don't want to display a notice.

File: client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	return useQuery< JetpackConnectionHealth, unknown, JetpackConnectionHealth >( {
+		queryKey: [ JETPACK_CONNECTION_HEALTH_QUERY_KEY, siteId ],
+		queryFn: () =>
+			wp.req.get( {
+				path: `/sites/${ siteId }/jetpack-connection-health`,
+				apiNamespace: 'wpcom/v2',
+			} ),
+		select: ( data ) => {
+			return data;
+		},
```
Comment (kozer) [inline client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts]: I don't think this function is necessary...

**Review by kozer — CHANGES_REQUESTED** [review]

File: client/components/jetpack/connection-health/index.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	return (
+		<Notice
+			status="is-warning"
+			showDismiss={ false }
+			text={ translate( 'Jetpack connection failed.' ) }
+		>
+			<NoticeAction href="#" external>
+				Attempt reconnection
+			</NoticeAction>
```
Comment (kozer) [inline client/components/jetpack/connection-health/index.tsx]: Can we, either:
- Remove that, as it doesn't do anything
- Make it point to the correct URL? ( `/wp-admin/admin.php?page=jetpack#/reconnect`)

**Review by kozer — CHANGES_REQUESTED** [review]

**Comment by kozer** [comment]
@wojtekn , I reviewed this ( although I understand that this is in progress ) and I left some comments.
Also I have updated the ticket [here](https://github.com/Automattic/dotcom-forge/issues/3202#issuecomment-1658134449) about designs. 
TL;DR: We' re good as is, expect we need to change the `info` to an `error` icon

File: client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts (hunk 2 above)
Comment (wojtekn) [reply client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts]: Thanks, I've removed it.

File: client/components/jetpack/connection-health/index.tsx (hunk 3 above)
Comment (wojtekn) [reply client/components/jetpack/connection-health/index.tsx]: I pointed it to the support document for now, and we can change it to the reconnect URL when we work on particular error cases. What do you think?

File: client/components/jetpack/connection-health/index.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		} );
+
+	if (
+		isLoadingJetpackConnectionHealth ||
+		isErrorCheckJetpackConnectionHealth ||
+		jetpackConnectionHealth?.is_healthy
+	) {
+		return;
+	}
+
```
Comment (danielbachhuber) [inline client/components/jetpack/connection-health/index.tsx:44]: Let's add a Tracks event to keep track of how often this is displayed

File: client/components/jetpack/connection-health/index.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+			showDismiss={ false }
+			text={ translate( 'Jetpack connection failed.' ) }
+		>
+			<NoticeAction
+				href={ localizeUrl(
+					'https://wordpress.com/support/why-is-my-site-down/#theres-an-issue-with-your-sites-jetpack-connection'
+				) }
+				external
+			>
+				Attempt reconnection
```
Comment (danielbachhuber) [inline client/components/jetpack/connection-health/index.tsx]: "Learn how to fix"? Also needs to be translated

Comment (kozer) [reply client/components/jetpack/connection-health/index.tsx:45]: I propose to add the following two events:
- **On load: (to track how many users see this message):** A proposed name is `calypso_connection_health_issue_view`
- **On "Learn how to fix" click ( to track how many users actually decided to investigate this further ):** A proposed name is `calypso_connection_health_issue_click`

cc: @wojtekn , @danielbachhuber

File: client/components/jetpack/connection-health/index.tsx (hunk 4 above)
Comment (kozer) [reply client/components/jetpack/connection-health/index.tsx:44]: I propose to add the following two events:
- **On load: (to track how many users see this message):** A proposed name is `calypso_connection_health_issue_view`
- **On "Learn how to fix" click ( to track how many users actually decided to investigate this further ):** A proposed name is `calypso_connection_health_issue_click`

cc: @wojtekn , @danielbachhuber

Comment (danielbachhuber) [reply client/components/jetpack/connection-health/index.tsx:44]: Sounds good!

File: client/components/jetpack/connection-health/index.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		jetpackConnectionHealth?.is_healthy
+	) {
+		return;
+	}
+
+	dispatch(
+		recordTracksEvent( 'calypso_jetpack_connection_health_issue_view', {
+			type: 'default',
+		} )
+	);
```
Comment (kozer) [inline client/components/jetpack/connection-health/index.tsx]: By doing that, when the user clicks on any tab, it will cause this to be triggered again, causing incorrect metrics.
I manage to fix it, by applying the following diff:
```
diff --git a/client/components/jetpack/connection-health/index.tsx b/client/components/jetpack/connection-health/index.tsx
index 0fcf37270b..8fda4bb9de 100644
--- a/client/components/jetpack/connection-health/index.tsx
+++ b/client/components/jetpack/connection-health/index.tsx
@@ -1,6 +1,6 @@
 import { localizeUrl } from '@automattic/i18n-utils';
 import { useTranslate } from 'i18n-calypso';
-import { useState } from 'react';
+import { useRef, useState } from 'react';
 import Notice from 'calypso/components/notice';
 import NoticeAction from 'calypso/components/notice/notice-action';
 import { useDispatch } from 'calypso/state';
@@ -13,6 +13,8 @@ interface Props {
 
 export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
 	const translate = useTranslate();
+	const isFirstRender = useRef( true );
+
 	const dispatch = useDispatch();
 
 	const [ isErrorCheckJetpackConnectionHealth, setIsErrorCheckJetpackConnectionHealth ] =
@@ -41,11 +43,14 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
 		return;
 	}
 
-	dispatch(
-		recordTracksEvent( 'calypso_jetpack_connection_health_issue_view', {
-			type: 'default',
-		} )
-	);
+	if ( isFirstRender.current ) {
+		isFirstRender.current = false;
+		dispatch(
+			recordTracksEvent( 'calypso_jetpack_connection_health_issue_view', {
+				type: 'default',
+			} )
+		);
+	}
 
 	return (
 		<Notice
```

**Review by kozer — CHANGES_REQUESTED** [review]

File: client/components/jetpack/connection-health/index.tsx (hunk 6 above)
Comment (wojtekn) [reply client/components/jetpack/connection-health/index.tsx]: Thanks, good point. As we talked on Slack, I will change it to use the `TrackComponentView` component which is a mechanism created precisely for that purpose.

**Review by kozer — APPROVED** [review]

File: client/components/jetpack/connection-health/index.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	return (
+		<>
+			<TrackComponentView
+				eventName="calypso_jetpack_connection_health_issue_view"
+				eventProperties={ { type: 'default' } }
+			/>
+			<Notice
+				status="is-error"
+				showDismiss={ false }
+				text={ translate( 'Jetpack connection failed.' ) }
```
Comment (danielbachhuber) [inline client/components/jetpack/connection-health/index.tsx]:
```suggestion
				text={ translate( 'Jetpack is unable to communicate with your site.' ) }
```

File: config/production.json [hunk 8]
```diff
@@ -174,7 +174,8 @@
 		"woop": false,
 		"wordpress-action-search": false,
 		"wpcom-user-bootstrap": true,
-		"yolo/hosting-metrics-i1": false
+		"yolo/hosting-metrics-i1": false,
+		"yolo/jetpack-error-ux-i1": false
```
Comment (danielbachhuber) [inline config/production.json]: I thought we were going to enable this for all users?

**Review by danielbachhuber — APPROVED** [review]
I added a fatal to my site and it seems to work well!


https://github.com/Automattic/wp-calypso/assets/36432/67fdaf5b-147f-460d-a199-c94168a7b961

I don't have any strong opinions on the React bits

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8679750

Thank you @wojtekn for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8679750).

File: client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+				apiNamespace: 'wpcom/v2',
+			} ),
+		meta: {
+			persist: false,
+		},
+		staleTime: 10 * 1000,
+		onError: options?.onError,
+		onSuccess: ( data ) => {
+			if ( data?.is_healthy ) {
+				dispatch( setJetpackConnectionHealthy( siteId ) );
```
Comment (flootr) [inline client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts:40]: Syncing Redux state from RQ is an anti-pattern and should be avoided if possible as it's an easy source for bugs. I came across this piece of code while working on the [RQv5](https://github.com/Automattic/wp-calypso/pull/84338) update. In v5 `on*` handlers are removed from the `useQuery` API (for already mentioned reasons) so we have to migrate all existing usages.

I created #84570 which refactors this code to `useEffect` to unblock the v5 upgrade although it's not ideal as we shouldn't be state-syncing at all. However, looking at the code I can see the Redux selector usages are not few and #81086 introduced complexity on top of that so it's not a straightforward refactor to exclusively use a RQ hook for connection health status.

@wojtekn @kozer @danielbachhuber would any of you be able to look into this and help unmangle usage of this query hook with the Redux state? Ideally we'd either use this query hook wherever we want to retrieve the state of the connection OR migrate to our legacy data layer and not use RQ at all if we want to keep the Redux stuff. I think the former is preferred in general but I'm not fully aware of all the context there may be for choosing Redux state in this case.

Comment (danielbachhuber) [reply client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts:40]: @flootr Yep, I think @kozer is the best person to help out at this point. I created an issue to track the work: https://github.com/Automattic/dotcom-forge/issues/4647

Comment (flootr) [reply client/components/jetpack/connection-health/use-check-jetpack-connection-health.ts:40]: Thank you!
